components/sup/src/manager/file_watcher.rs

```diff
@@ -1,3464 +1,2652 @@
-            fs::read_dir(&path).unwrap_or_else(|err| {
-                                   panic!("failed to read directory {}: {}, debug info:\n{}",
-                                          path.display(),
-                                          err,
-                                          self.debug_info,)
-                               })
-                               .map(|rde| {
-                                   rde.unwrap_or_else(|err| {
-                                          panic!("failed to get entry for {}: {}, debug info:\n{}",
-                                                 path.display(),
-                                                 err,
-                                                 self.debug_info,)
-                                      })
-                                      .path()
-                               })
-                               .collect()
-        }
+                assert_eq!(1, fw.callbacks.callback_type_count(Appeared));
+                assert_eq!(1, fw.callbacks.path_instance_count(&td_watched_file));
 
-        fn parent_is_watched(&self, path: &Path) -> bool {
-            self.path_is_watched(&self.get_parent(path))
-        }
+                Ok(())
+            }
 
-        fn path_is_watched(&self, path: &Path) -> bool {
-            match self.watched_dirs {
-                Some(wd) => wd.contains(path),
-                None => false,
+            #[test]
+            fn move_symlink_to_enclosing_directory_pw() -> Result<()> {
+                let lock = lock_env_var();
+                lock.set("aarch64-darwin");
+                let r = move_symlink_to_enclosing_directory();
+                lock.unset();
+                r
             }
         }
+    }
 
-        fn get_parent(&self, path: &Path) -> PathBuf {
-            path.parent()
-                .unwrap_or_else(|| {
-                    panic!("path {} has no parent, debug info:\n{}",
-                           path.display(),
-                           self.debug_info)
-                })
-                .to_owned()
-        }
+    mod scenarios {
+        use super::*;
 
-        fn prepend_root(&self, p: &Path) -> PathBuf {
-            prepend_root_impl(self.root, p, self.debug_info)
-        }
-    }
+        #[test]
+        #[allow(clippy::cognitive_complexity)]
+        // This unit test approximates an issue discovered via manual testing on Windows.
+        // * deploy a service in habitat
+        // * see that the user.toml is present and in effect
+        // * rename the user.toml file to user.tuml
+        // * see that Habitat begins to use the default configuration
+        // * delete the config directory that user.toml (would) live in
+        // * recreate the config directory
+        // * recreate the user.toml file
+        // The result was that the UserConfigWatcher was not seeing the user.toml when it returned
+        fn first() -> Result<()> {
```

Comment:
did you manage to get this to fail before your fix? I remember it was passing with the bug unless you made some changes. If it never failed, then I would remove the test.

Reply:
We talked through things and this was failing when I first wrote it and I iterated until it wasn't so we're keeping it.